build-common/src/org/jetbrains/kotlin/incremental/KotlinClassInfo.kt

```diff
@@ -84,123 +109,148 @@ class KotlinClassInfo constructor(
+        }
     }
+}
 
-    override fun visitMethod(access: Int, name: String, desc: String, signature: String?, exceptions: Array<out String>?): MethodVisitor? {
-        // Note: Do not filter out private methods here because a *public* inline function may actually have a *private* corresponding JVM
-        // method in the bytecode (see `InlineOnlyKt.isInlineOnlyPrivateInBytecode`).
-        // Just filter the methods based on the given `inlineFunctionsAndAccessors` set.
-        val method = JvmMemberSignature.Method(name, desc)
-        if (method !in inlineFunctionsAndAccessors) return null
+/** Computes the snapshot of a Java class represented by a [ClassNode]. */
+object ClassNodeSnapshotter {
 
+    fun snapshotClass(classNode: ClassNode): Long {
         val classWriter = ClassWriter(0)
+        classNode.accept(classWriter)
+        return classWriter.toByteArray().hashToLong()
+    }
 
-        // The `version` and `name` parameters are important (see KT-38857), the others can be null.
-        classWriter.visit(/* version */ classVersion!!, /* access */ 0, /* name */ "ClassWithOneMethod", null, null, null)
-
-        return object : MethodVisitor(Opcodes.API_VERSION, classWriter.visitMethod(access, name, desc, signature, exceptions)) {
-            override fun visitEnd() {
-                result[method] = classWriter.toByteArray().md5()
-            }
+    fun snapshotClassExcludingMembers(classNode: ClassNode, alsoExcludeKotlinMetaData: Boolean = false): Long {
+        val originalFields = classNode.fields
+        val originalMethods = classNode.methods
+        val originalVisibleAnnotations = classNode.visibleAnnotations
+        classNode.fields = emptyList()
+        classNode.methods = emptyList()
+        if (alsoExcludeKotlinMetaData) {
+            classNode.visibleAnnotations = originalVisibleAnnotations?.filterNot { it.desc == "Lkotlin/Metadata;" }
         }
+        return snapshotClass(classNode).also {
+            classNode.fields = originalFields
+            classNode.methods = originalMethods
+            classNode.visibleAnnotations = originalVisibleAnnotations
+        }
+    }
+
+    fun snapshotField(fieldNode: FieldNode): Long {
+        val classNode = emptyClass()
+        classNode.fields.add(fieldNode)
+        return snapshotClass(classNode)
+    }
+
+    fun snapshotMethod(methodNode: MethodNode, classVersion: Int): Long {
+        val classNode = emptyClass()
+        classNode.version = classVersion // Class version is required for method bodies (see KT-38857)
+        classNode.methods.add(methodNode)
+        return snapshotClass(classNode)
+    }
+
+    /**
+     * Sorts fields and methods in the given class.
+     *
+     * This is useful when we want to ensure a change in the order of the fields and methods doesn't impact the snapshot (i.e., if their
+     * order has changed in the `.class` file, it shouldn't require recompilation of the other source files).
+     */
```

Comment:
Future improvement: This seems to be the wrong place to do this. A more robust approach is to sort these entries in the snapshot.

Reply:
As discussed offline, your suggestions was to sort the fields/methods directly when creating their snapshots:
```
sortClassMembers(classNode)
...
val constantSnapshots: Map<String, Long> = classNode.fields // Your suggestion: Sort them here
    .filter { !it.isPrivate() && it.isStaticFinal() }
    .associate { it.name to snapshotField(it) }
```        

but it's fine to sort them in a separate earlier step as done in the current code.